### PR TITLE
Feature/pyqtchart integration

### PR DESCRIPTION
**Pull request by bradyzp**
This branch contains the PyQtGraph integrations, some of the work here is deprecated due to our decision to move away from using PyQtGraph graphic flowcharts for the transformation workflow.

ToDo: Clean-up deprecated code after merging in the new transformation branch. Specifically features in the dgp/lib/transform package which will be superseded by the new transform workflows.

Features Added in this Branch:
- New Flight-Line selection plot based on PyQtGraph - this provides a much more responsive UI experience when viewing raw data to select flight lines.
- Flight Workspaces (dgp/gui/workspaces): A modular approach to handling the sub-tab widgets for each flight - allowing fast switching between multiple open flights and in future the ability to view & mutate data for each flight in its own tab. Workspaces can be dynamically loaded, and each is contained within its own python module.
- Added an excepthook method in the program __main__ module, and in the test runner config which captures and prints stack-traces for uncaught Qt exceptions, which otherwise crash with an integer error value.

### CODE REVIEW

File: dgp/gui/dialogs.py [hunk 1]
```diff
@@ -416,9 +414,11 @@ def path(self, value):
             self.line_path.setText('None')
             return
 
+        print("Raw path value: ", value)
```
Comment (cbertinato) [inline dgp/gui/dialogs.py]: For debug, or still need this?

File: dgp/gui/main.py [hunk 2]
```diff
@@ -234,37 +215,39 @@ def _launch_tab(self, index: QtCore.QModelIndex=None, flight=None) -> None:
...
-            self._context_tree.setModel(None)
+            # self._context_tree.setModel(None)
             return
-        tab = self._tabs.widget(index)  # type: FlightTab
-        self._context_tree.setModel(tab.context_model)
-        self._context_tree.expandAll()
+        tab = self._flight_tabs.widget(index)  # type: FlightTab
+        if tab.subtab_widget():
+            print("Active flight subtab has a widget")
+        # self._context_tree.setModel(tab.context_model)
```
Comment (cbertinato) [inline dgp/gui/main.py]: Can we remove these commented lines? Also below at 249 and 250?

File: dgp/gui/plotting/mplutils.py [hunk 3]
```diff
@@ -380,6 +390,95 @@ def reset_view(self, x_margin=None, y_margin=None):
...
+        for group in self.groups:
+            edge = group.get_edge(xdata, inner=False)
+            if edge in ('left', 'right'):
+
+                edge_grp = group
+                self.parent.setCursor(QtCore.Qt.SizeHorCursor)
+                group.set_edge(edge, 'red', select=False)
+                break
+        else:
+            # group.set_edge('', 'black', select=False)
```
Comment (cbertinato) [inline dgp/gui/plotting/mplutils.py]: Still need this commented code?

File: dgp/gui/plotting/plotters.py [hunk 4]
```diff
@@ -1,45 +1,326 @@
...
+from .flightregion import LinearFlightRegion
 
+import pyqtgraph as pg
+from pyqtgraph.graphicsItems.LinearRegionItem import LinearRegionItem
 
 _log = logging.getLogger(__name__)
+
+
+"""
+TODO: Many of the classes here are not used, in favor of the PyQtGraph line selection interface.
```
Comment (cbertinato) [inline dgp/gui/plotting/plotters.py:46]: I think it's ok to leave this around until we work out the remaining issues with the new plotting interface.

File: dgp/gui/views.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        raise NotImplementedError("Remove data not yet implemented.")
+        # Confirmation Dialog
+        confirm = QtWidgets.QMessageBox(parent=self.parent())
+        confirm.setStandardButtons(QtWidgets.QMessageBox.Ok)
+        confirm.setText("Are you sure you wish to delete: {}".format(item.filename))
+        confirm.setIcon(QtWidgets.QMessageBox.Question)
+        confirm.setWindowTitle("Confirm Delete")
+        res = confirm.exec_()
+        if res:
+            print("Emitting item_removed signal")
```
Comment (cbertinato) [inline dgp/gui/views.py]: Should this print statements be moved into the logs?

File: dgp/gui/workspace.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        self._plot_tab = PlotTab(label="Plot", flight=flight, axes=3)
+        self._workspace.addTab(self._plot_tab, "Plot")
+
+        self._transform_tab = TransformTab("Transforms", flight)
+        self._workspace.addTab(self._transform_tab, "Transforms")
+
+        self._line_proc_tab = LineProcessTab("Line Processing", flight)
+        self._workspace.addTab(self._line_proc_tab, "Line Processing")
+
+        # self._map_tab = WorkspaceWidget("Map")
```
Comment (cbertinato) [inline dgp/gui/workspace.py]: Can this be removed?

File: examples/pyqtgraph_line_selection_plot.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+class PlotExample(QtWidgets.QMainWindow):
+    def __init__(self):
+        super().__init__()
+        self.setWindowTitle('Plotter Testing')
+        self.setBaseSize(Qt.QSize(600, 600))
+        self._flight = project.Flight(None, 'test')
+
+        self._plot = LineSelectPlot(flight=self._flight, rows=3)
+        self._plot.line_changed.connect(lambda upd: print(upd))
+        # self.plot.figure.canvas.mpl_connect('pick_event', lambda x: print(
```
Comment (cbertinato) [inline examples/pyqtgraph_line_selection_plot.py]: Can these be removed?

File: examples/pyqtgraph_line_selection_plot.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+                            periods=1000)
+        ser = Series([np.sin(x)*3 for x in np.arange(0, 100, 0.1)], index=idx)
+        p0 = self._plot.plots[0]
+        p0.add_series(ser)
+        print("new xlim: ", p0.get_xlim())
+        x0, x1 = p0.get_xlim()
+        xrng = x1 - x0
+        tenpct = xrng * .1
+
+        # lri = LinearRegionItem()
```
Comment (cbertinato) [inline examples/pyqtgraph_line_selection_plot.py]: And here?

**Review by cbertinato — CHANGES_REQUESTED** [review]
Nice work, Zach! I haven't done any QA testing yet, but if anything arises when I do, then I'll just make an issue.

File: dgp/gui/plotting/mplutils.py (hunk 3 above)
Comment (bradyzp) [reply dgp/gui/plotting/mplutils.py]: Nope, cleaned that out. In fact I'm not sure about the future of the mplutils classes, they might be useful for plotting transformed data, but like below with plotters.py will have to consider whether to keep them around.

File: dgp/gui/views.py (hunk 5 above)
Comment (bradyzp) [reply dgp/gui/views.py]: Removed those as I was just using them to debug something there, turned the exception handler into a logging statement.

**Review by cbertinato — DISMISSED** [review]

**Review by cbertinato — APPROVED** [review]

**Pull request merged**